### ISSUE TITLE
Spam Detection

### ISSUE THREAD

**Issue by ahmedazeemicloud44-sketch** [ahmedazeemicloud44-sketch/gdp-dashboard#1]
import pandas as pd
import nltk
import string
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords

data = {
    "label": ["ham", "spam", "ham", "spam", "ham", "spam", "ham", "spam", "ham", "spam"],
    "message": [
        "Hey, are we still meeting for lunch today?",
        "Congratulations! You've won a $1000 Walmart gift card. Click here to claim.",
        "Don't forget to bring your notes to class.",
        "URGENT! Your account has been compromised. Verify now!",
        "Can you call me when you get a chance?",
        "Free entry in 2 a weekly competition to win FA Cup tickets!",
        "Happy birthday! Have a great day!",
        "Winner! You have been selected for a free cruise. Reply now!",
        "Let's watch a movie this weekend.",
        "You have won a free iPhone! Claim your prize."
    ]
}

# Convert to DataFrame
df = pd.DataFrame(data)
print(df)

nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')

# Initialize lemmatizer and stopwords
lemmatizer = WordNetLemmatizer()
stop_words = set(stopwords.words('english'))

def preprocess_text(text):
    # 1. Lowercase
    text = text.lower()

    # 2. Remove punctuation
    text = text.translate(str.maketrans('', '', string.punctuation))

    # 3. Tokenize
    words = nltk.word_tokenize(text)

    # 4. Remove stopwords & 5. Lemmatize
    words = [lemmatizer.lemmatize(word) for word in words if word not in stop_words]

    # Join back into a string
    return ' '.join(words)

# Apply preprocessing to the 'message' column
df['clean_text'] = df['message'].apply(preprocess_text)
print(df[['message', 'clean_text']])
from sklearn.feature_extraction.text import TfidfVectorizer
# You can customize max_features to limit the number of words
vectorizer = TfidfVectorizer(max_features=50)  # For small dataset, 50 features are enough
# 'clean_text' column contains our preprocessed text
X_tfidf = vectorizer.fit_transform(df['clean_text'])
# Convert sparse matrix to array (optional)
X_array = X_tfidf.toarray()

print("TF-IDF Feature Names:", vectorizer.get_feature_names_out())
print("TF-IDF Vectors:\n", X_array)
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
y = le.fit_transform(df['label']) # Corrected 'labels' to 'df['label']'

print(y)  # Output: [0 1 0 1] -> 0=ham, 1=spam

from sklearn.model_selection import train_test_split

# X_tfidf is your features, y is the labels
X_train, X_test, y_train, y_test = train_test_split(X_tfidf, df['label'], test_size=0.2, random_state=42)

print("Training samples:", X_train.shape[0])
print("Testing samples:", X_test.shape[0])

from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report

# Initialize model
model = MultinomialNB()

# Train model
model.fit(X_train, y_train)

# Predict on test data
y_pred = model.predict(X_test)

# Evaluate
print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

import joblib

joblib.dump(model, "spam_model.pkl")
joblib.dump(vectorizer, "tfidf_vectorizer.pkl")
